File: backend/src/services/auth/role.py
from sqlalchemy.orm import Session
from typing import List
from fastapi import HTTPException, status
from src.models.role import Role
from src.models.permission import Permission
from src.schemas.auth import RoleCreate, RoleUpdate, RoleResponse
from src.services.auth.permissions import Permissions


def list_roles(db: Session) -> List[RoleResponse]:
    roles = db.query(Role).all()
    return [
        RoleResponse(id=r.id, name=r.name, permissions=[p.name for p in r.permissions])
        for r in roles
    ]


def create_role(db: Session, role: RoleCreate) -> RoleResponse:
    if db.query(Role).filter(Role.name == role.name).first():
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="Role name already exists"
        )
    db_role = Role(name=role.name)
    if role.permissions:
        unique_permissions = list(set(role.permissions))
        valid_permissions = set(Permissions.all())
        invalid = [p for p in unique_permissions if p not in valid_permissions]
        if invalid:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid permissions: {', '.join(invalid)}",
            )
        perms = db.query(Permission).filter(Permission.name.in_(unique_permissions)).all()
        if len(perms) != len(unique_permissions):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="One or more permissions do not exist in the database",
            )
        db_role.permissions = perms
    db.add(db_role)
    db.commit()
    db.refresh(db_role)
    return RoleResponse(
        id=db_role.id,
        name=db_role.name,
        permissions=[p.name for p in db_role.permissions],
    )


def get_role(db: Session, role_id: int) -> RoleResponse:
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Role not found"
        )
    return RoleResponse(
        id=role.id, name=role.name, permissions=[p.name for p in role.permissions]
    )


def update_role(db: Session, role_id: int, update: RoleUpdate) -> RoleResponse:
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Role not found"
        )
    if (
        update.name != role.name
        and db.query(Role).filter(Role.name == update.name, Role.id != role_id).first()
    ):
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="Role name already exists"
        )
    role.name = update.name
    if update.permissions is not None:
        unique_permissions = list(set(update.permissions))
        valid_permissions = set(Permissions.all())
        invalid = [p for p in unique_permissions if p not in valid_permissions]
        if invalid:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid permissions: {', '.join(invalid)}",
            )
        perms = (
            db.query(Permission).filter(Permission.name.in_(unique_permissions)).all()
        )
        if len(perms) != len(unique_permissions):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="One or more permissions do not exist in the database",
            )
        role.permissions = perms
    db.commit()
    db.refresh(role)
    return RoleResponse(
        id=role.id, name=role.name, permissions=[p.name for p in role.permissions]
    )


def delete_role(db: Session, role_id: int) -> None:
    role = db.query(Role).filter(Role.id == role_id).first()
    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Role not found"
        )
    db.delete(role)
    db.commit()

File: backend/src/services/auth/role.rs
use sqlx::PgPool;

use crate::errors::AppError;
use crate::models::role::Role;
use crate::models::permission::Permission;
use crate::schemas::auth::{RoleResponse, PermissionResponse};

pub async fn get_all_roles(pool: &PgPool) -> Result<Vec<RoleResponse>, AppError> {
    let roles = sqlx::query_as::<_, Role>("SELECT * FROM roles ORDER BY id")
        .fetch_all(pool)
        .await
        .map_err(AppError::Database)?;

    let mut responses = Vec::with_capacity(roles.len());
    for role in roles {
        let perms = get_role_permissions(pool, role.id).await?;
        responses.push(RoleResponse {
            id: role.id,
            name: role.name,
            permissions: perms,
        });
    }
    Ok(responses)
}

pub async fn get_role_by_id(pool: &PgPool, role_id: i32) -> Result<RoleResponse, AppError> {
    let role = sqlx::query_as::<_, Role>("SELECT * FROM roles WHERE id = $1")
        .bind(role_id)
        .fetch_one(pool)
        .await
        .map_err(|e| match e {
            sqlx::Error::RowNotFound => AppError::NotFound("Role".to_string(), role_id.to_string()),
            e => AppError::Database(e),
        })?;
    let perms = get_role_permissions(pool, role.id).await?;
    Ok(RoleResponse {
        id: role.id,
        name: role.name,
        permissions: perms,
    })
}

pub async fn create_role(pool: &PgPool, name: &str) -> Result<RoleResponse, AppError> {
    let role = sqlx::query_as::<_, Role>("INSERT INTO roles (name) VALUES ($1) RETURNING *")
        .bind(name)
        .fetch_one(pool)
        .await
        .map_err(|e| match e {
            sqlx::Error::Database(ref db_err)
                if db_err.constraint() == Some("roles_name_key") =>
            {
                AppError::Conflict(format!("Role '{}' already exists", name))
            }
            e => AppError::Database(e),
        })?;
    Ok(RoleResponse {
        id: role.id,
        name: role.name,
        permissions: vec![],
    })
}

pub async fn assign_permission_to_role(
    pool: &PgPool,
    role_id: i32,
    permission_id: i32,
) -> Result<(), AppError> {
    sqlx::query(
        "INSERT INTO role_permissions (role_id, permission_id) VALUES ($1, $2) ON CONFLICT DO NOTHING",
    )
    .bind(role_id)
    .bind(permission_id)
    .execute(pool)
    .await
    .map_err(AppError::Database)?;
    Ok(())
}

pub async fn remove_permission_from_role(
    pool: &PgPool,
    role_id: i32,
    permission_id: i32,
) -> Result<(), AppError> {
    sqlx::query("DELETE FROM role_permissions WHERE role_id = $1 AND permission_id = $2")
        .bind(role_id)
        .bind(permission_id)
        .execute(pool)
        .await
        .map_err(AppError::Database)?;
    Ok(())
}

pub async fn assign_role_to_user(
    pool: &PgPool,
    user_id: i32,
    role_id: i32,
) -> Result<(), AppError> {
    sqlx::query(
        "INSERT INTO user_roles (user_id, role_id) VALUES ($1, $2) ON CONFLICT DO NOTHING",
    )
    .bind(user_id)
    .bind(role_id)
    .execute(pool)
    .await
    .map_err(AppError::Database)?;
    Ok(())
}

pub async fn remove_role_from_user(
    pool: &PgPool,
    user_id: i32,
    role_id: i32,
) -> Result<(), AppError> {
    sqlx::query("DELETE FROM user_roles WHERE user_id = $1 AND role_id = $2")
        .bind(user_id)
        .bind(role_id)
        .execute(pool)
        .await
        .map_err(AppError::Database)?;
    Ok(())
}

pub async fn get_user_roles(pool: &PgPool, user_id: i32) -> Result<Vec<RoleResponse>, AppError> {
    let roles = sqlx::query_as::<_, Role>(
        r#"
        SELECT r.* FROM roles r
        JOIN user_roles ur ON ur.role_id = r.id
        WHERE ur.user_id = $1
        ORDER BY r.id
        "#,
    )
    .bind(user_id)
    .fetch_all(pool)
    .await
    .map_err(AppError::Database)?;

    let mut responses = Vec::with_capacity(roles.len());
    for role in roles {
        let perms = get_role_permissions(pool, role.id).await?;
        responses.push(RoleResponse {
            id: role.id,
            name: role.name,
            permissions: perms,
        });
    }
    Ok(responses)
}

async fn get_role_permissions(pool: &PgPool, role_id: i32) -> Result<Vec<PermissionResponse>, AppError> {
    let perms = sqlx::query_as::<_, Permission>(
        r#"
        SELECT p.* FROM permissions p
        JOIN role_permissions rp ON rp.permission_id = p.id
        WHERE rp.role_id = $1
        ORDER BY p.id
        "#,
    )
    .bind(role_id)
    .fetch_all(pool)
    .await
    .map_err(AppError::Database)?;
    Ok(perms.into_iter().map(|p| PermissionResponse { id: p.id, name: p.name }).collect())
}
